<launch>
  <!-- Set this to your camera's name -->
  <arg name="cam_name" value="jetson_cam" />
  <arg name="gui" default="true"/>
  <arg name="feed" default="1"/>
  <arg name="side" default="orange"/>

  <!-- Start the GSCAM node -->
  <node pkg="gscam" type="gscam" name="$(arg cam_name)">
    <param name="camera_name" value="$(arg cam_name)" />
    <param name="camera_info_url" value="file:///home/nvidia/.camera_info/camera_calibration.yaml" />
    <param name="gscam_config" value="nvcamerasrc ! video/x-raw(memory:NVMM), width=(int)1280, height=(int)720,format=(string)I420, framerate=(fraction)20/1 ! nvvidconv flip-method=0 ! video/x-raw, format=(string)BGRx ! videoconvert ! video/x-raw, format=(string)BGR" />
    

    <remap from="camera/image_raw" to="$(arg cam_name)/image_raw" />
    <remap from="set_camera_info" to="$(arg cam_name)/set_camera_info" />
  </node>

  <!-- Provide rectification -->
  <node pkg="image_proc" type="image_proc" name="jetson_image_proc"
        ns="$(arg cam_name)" />

   <!-- View the raw and rectified output -->
   <!-- <node pkg="image_view" type="image_view" name="image_raw_feed" >       --> 
   <!--   <remap from="image" to="/$(arg cam_name)/image_raw" />               -->
   <!-- </node>                                                                -->
               
   <!-- <node pkg="image_view" type="image_view" name="iamge_rect_feed" >      -->
   <!--   <remap from="image" to="/$(arg cam_name)/image_rect_color" />        -->
   <!-- </node>                                                                -->

  <arg name="launch_prefix" default="" /> <!-- set to value="gdbserver localhost:10000" for remote debugging -->
  <arg name="node_namespace" default="apriltags2_tag_detection_node" />
  <arg name="camera_frame" default="camera" />
  <arg name="image_topic" default="image_rect" />


  <!-- Here be Apriltags2 code -->
  <!-- Set parameters -->
  <rosparam command="load" file="$(find eurobot_localisation)/config/settings.yaml" ns="$(arg node_namespace)" />
  <rosparam command="load" file="$(find eurobot_localisation)/config/tags.yaml" ns="$(arg node_namespace)" />
  
  <node pkg="apriltags2_ros" type="apriltags2_ros_continuous_node" name="$(arg node_namespace)" clear_params="true" output="screen" launch-prefix="$(arg launch_prefix)" >
    <!-- Remap topics from those used in code to those on the ROS network -->
    <remap from="image_rect" to="$(arg cam_name)/$(arg image_topic)" />
    <remap from="camera_info" to="$(arg cam_name)/camera_info" />

    <param name="camera_frame" type="str" value="$(arg camera_frame)" />
    <param name="publish_tag_detections_image" type="bool" value="true" />      <!-- default: false -->
  </node>

  <node pkg="eurobot_localisation" type="eurobot_localisation_node" name="visual_pose_estimator" >
    <param name="colour" type="str" value="$(arg side)" />
  </node>
  <node pkg="eurobot_localisation" type="eurobot_tf_broadcaster_node" name="eurobot_arena_tf_generator" >
    <param name="colour" type="str" value="$(arg side)" />
  </node>

  <node if="$(arg feed)" pkg="image_view" type="image_view" name="tag_detection_feed" >     
    <remap from="image" to="/tag_detections_image" />       
  </node>        
    

  <node if="$(arg gui)" pkg="rviz" type="rviz" name="tf_visualiser" output="screen" args="-d $(find eurobot_localisation)/config/eurobot_localisation.rviz"/>

</launch>
